File: examples/sample-java/src/main/java/com/monetai/sample/java/FakeProductAdapter.java
package com.monetai.sample.java;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class FakeProductAdapter extends RecyclerView.Adapter<FakeProductAdapter.ProductViewHolder> {

    public interface OnVisibleListener {
        void onVisible(FakeProduct product);
    }

    private final OnVisibleListener onVisibleListener;
    private List<FakeProduct> items = new ArrayList<>();

    public FakeProductAdapter(OnVisibleListener onVisibleListener) {
        this.onVisibleListener = onVisibleListener;
    }

    public void submitList(List<FakeProduct> newItems) {
        this.items = newItems;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_fake_product, parent, false);
        return new ProductViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        FakeProduct item = items.get(position);
        holder.bind(item);
        if (onVisibleListener != null) {
            onVisibleListener.onVisible(item);
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    static class ProductViewHolder extends RecyclerView.ViewHolder {
        private final TextView title;
        private final TextView description;
        private final TextView price;
        private final TextView regularPrice;

        ProductViewHolder(@NonNull View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.textProductTitle);
            description = itemView.findViewById(R.id.textProductDescription);
            price = itemView.findViewById(R.id.textProductPrice);
            regularPrice = itemView.findViewById(R.id.textProductRegularPrice);
        }

        void bind(FakeProduct item) {
            title.setText(item.title);
            description.setText(item.description);
            price.setText(item.currencyCode + " " + String.format("%.2f", item.price));
            regularPrice.setText(item.currencyCode + " " + String.format("%.2f", item.regularPrice));
        }
    }
}

File: examples/sample-kotlin/src/main/java/com/monetai/sample/kotlin/FakeProductAdapter.kt
package com.monetai.sample.kotlin

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class FakeProductAdapter(
    private val onVisible: (FakeProduct) -> Unit
) : RecyclerView.Adapter<FakeProductAdapter.ProductViewHolder>() {
    
    private var items: List<FakeProduct> = emptyList()
    
    fun submitList(newItems: List<FakeProduct>) {
        items = newItems
        notifyDataSetChanged()
    }
    
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_fake_product, parent, false)
        return ProductViewHolder(view)
    }
    
    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) {
        val item = items[position]
        holder.bind(item)
        onVisible(item)
    }
    
    override fun getItemCount(): Int = items.size
    
    class ProductViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val title: TextView = itemView.findViewById(R.id.textProductTitle)
        private val description: TextView = itemView.findViewById(R.id.textProductDescription)
        private val price: TextView = itemView.findViewById(R.id.textProductPrice)
        private val regularPrice: TextView = itemView.findViewById(R.id.textProductRegularPrice)
        
        fun bind(item: FakeProduct) {
            title.text = item.title
            description.text = item.description
            price.text = "${item.currencyCode} ${"%.2f".format(item.price)}"
            regularPrice.text = "${item.currencyCode} ${"%.2f".format(item.regularPrice)}"
        }
    }
}

